rfcs/0051-limit-zkapp-cmds-per-block.md

```diff
@@ -0,0 +1,156 @@
+## Summary
+
+During the ITN stress testing it was noticed that daemon's memory
+consumption tends to increase dramatically after a block containing a
+large number of zkApp commands. Before appropriate optimizations can
+be developed, we need a temporary solution to prevent nodes crashing
+due to insufficient memory. The idea is to limit the number of zkApp
+commands that can be included in any single block.
+
+## Motivation
+
+By limiting the number of zkApp commands going into blocks we avoid
+the aforementioned issue until a proper solution can be devised and
+implemented. The root cause of the issue is that proofs contained
+within these commands are stored in the scan state and tend to occupy
+a lot of space. Fixing these storage issues won't affect the
+protocol, so ideally we want a workaround that doesn't affect the
+protocol either, so that at the convenient time we can turn it off
+without making a fork.
+
+## Detailed design
+
+Since the solution should not affect the protocol, it should be
+implemented at the mempool/block producer boundary. In the mempool
+there is `transactions` function, which returns a sequence of
+transactions from the mempool in the order of decreasing transaction
+fees. The block producer module just takes a number of transactions
+from that sequence and includes them. It is relatively easy to limit
+the number of zkApp commands returned by this functions.  ZkApp
+commands above a certain limit would simply be discarded (from the
+sequence, but not from the mempool).
```

Comment:
I don't think this should be the layer to limit zkApp transactions. 
In the case where a zkApp transaction's fee payer fails (checked in `create_diff` function in staged_ledger.ml), the actual # of zkApp transactions in a block will be less than the limit, even if there were more in the mempool.

One option would be to let `transactions` continue to return the sequence as is and instead check in the `create_diff` function- the part where it validates transactions from the mempool. Any zkApp transaction after a certain number can be simply discarded. This will follow the same flow as if a transaction was invalid and ensures BP always fills a block up to the limit. 
With this approach you' would also know when a transaction was rejected due to the limit (vs being invalid) and so calling that out in rejected transaction logs will be useful info for debugging and maybe for node operators as well.

Reply:
I suppose it's doable too. Okay.

Reply:
What do we do in the case where discarding a zkapp transaction invalidates other txns in the sequence of applicable transactions (due to the necessary fee payer nonce increase not taking place)? I think this opens up a transaction pool DoS vector, similar to the one that we fixed with the 2 phase ledger apply change (although less severe). An attacker can enqueur `soft_limit+1` zkapps in the pool from the same fee payer, and then fill up the pool with transactions that extend that sequence. Unlike the other DoS vulnerability, they will likely end up having to pay for all these txns eventually, but if they are smart about rotating fee payers, they may be able to continue this attack by kicking their prior enqueued sequence out of the pool before the next block.

This could be considered acceptable, as they need to pay for at least `soft_limit` txns each time, and the attack is more expensive to execute each block, but it allows attackers to raise the fees of txns potentially without paying for every txn they add to the mempool.

Reply:
Well, even if there's no attack, it just so happens that only a zkApp command is skipped from a block, then further signed commands can still be executed, even though they're out of order. This can lead to lost zkApp commands, which become skipped at block `n` and then invalid at block `n+1` (because of the nonce increase due to a signed comamnd being included). This means we should oindeed implement the change in the mempool, so that all commands depending on the skipped one also get skipped.

Reply:
I'm not sure the solution should live at the mempool layer still. That requires you to ensure there is only ever `soft_limit`applicable zkapps within the first `hard_limit` applicable txns in the pool. I'm not sure how you would decide which fee payer sequences you would keep or discard when the `soft_limit` zkapps rule is violated (at least I can't think of a way that is both decidable and not susceptible to adversaries expunging other txns from the pool).

Reply:
> Well, even if there's no attack, it just so happens that only a zkApp command is skipped from a block, then further signed commands can still be executed, even though they're out of order.

To be clear, the attack here is that, if further zkapps cannot be included which are dependencies of the remaining signed commands in the pool, none of the remaining signed commands can be included either. That's why you need at least `soft_limit+1` zkapp commands. The attack also applies even if you just enqueue all the commands as zkapp commands, though that increases the weight of the txn, and thus increases the required fee to enqueue such commands, so I modeled this as though the remaining commands are signed.

Reply:
This requirement that mempool only ever holds applicable transactions is bad and should be relaxed anyway. I was working on a solution to allow this, but had to shelve it before I finished. It causes the network to loose valid transactions only because they arrive out-of-order. Here we encounter a similar issue: we allow ourselves to lose transactions only because they cannot be included right away. Already bad user experience will deteriorate further. I think the transactions depending on those skipped should be skipped too, but they should be allowed to remain in mempool, because they eventually will become valid.

Reply:
I don't quite agree, but regardless, that's a much larger change though and is certainly out of scope for getting this change into the upcoming Berkeley release. It would also require a separate RFC to illustrate how it is safe from DoS attacks.

This isn't the correct forum to discuss, but I believe there are other solutions to this problem that will fix the transaction ordering issue in a way similar to how other blockchains manage the problem.

Reply:
In other words, it's fine to drop zkApps that don't fit in the limit if there are signed commands depending on them? Do I understand correctly?

Reply:
Yep, that sounds accurate to me. Thanks Marcin.